gustaf/spline/ffd.py

```diff
@@ -0,0 +1,488 @@
+
+    def elevate_degree(self, *args, **kwargs):
+        """
+        Wrapper for Spline.elevate_degree
+
+        Parameters
+        -----------
+        *args:
+        **kwargs:
+
+        Returns
+        --------
+        None
+        """
+        self._spline.elevate_degree(*args, **kwargs)
+
+    def insert_knots(self, parametric_dimension, knots):
+        """
+        Wrapper for Spline.insert_knots
+
+        Parameters
+        -----------
+        *args:
+        **kwargs:
+
+        Returns
+        --------
+        None
+        """
+        if "knot_vectors" in self._spline.required_properties:
+            raise NotImplementedError(
+                "Can not perform knot insertion on Bezier spline.")
+        # scale knots from the original bounds into the new bounds
+        dim_bounds = self._orig_para_dim_ranges[parametric_dimension]
+        knots = (
+            (np.array(knots)-dim_bounds[0])
+            / (dim_bounds[-1]-dim_bounds[0])
+        )
+        self._spline.insert_knots(parametric_dimension, knots)
+
+    def remove_knots(self, parametric_dimension, knots, tolerance=1e-8):
+        """
+        Wrapper for Spline.remove_knots
+
+        Parameters
+        -----------
+        *args:
+        **kwargs:
+
+        Returns
+        --------
+        None
+        """
+        if "knot_vectors" in self._spline.required_properties:
+            raise NotImplementedError(
+                "Can not perform knot insertion on Bezier spline.")
+        # scale knots from the original bounds into the new bounds
+        dim_bounds = self._orig_para_dim_ranges[parametric_dimension]
+        knots = (
+            (np.array(knots)-dim_bounds[0])
+            / (dim_bounds[-1]-dim_bounds[0])
+        )
+        self._spline.remove_knots(
+            parametric_dimension, knots, tolerance=tolerance)
+        self._is_calculated = False
+
+    def reduce_degree(self, *args, **kwargs):
+        """
+        Wrapper for Spline.reduce_degree
+
+        Parameters
+        -----------
+        *args:
+        **kwargs:
+
+        Returns
+        --------
+        None
+        """
+        self._spline.reduce_degree(*args, **kwargs)
+        self._is_calculated = False
+
+    # The '*' prevents the title to be given as a positional argument
+    def show(
+            self, *, return_showable: bool = False,
+            return_discrete: bool = False,
+            title: str = "gustaf - FFD", **kwargs) -> Any:
+        """
+        Visualize. Shows the deformed mesh and the current spline. Currently
+        visualization is limited to vedo.
+
+        Parameters
+        ----------
+        title: str
+            Title of the vedo window. Defaults to "gustaf - FFD".
+        return_showable: bool
+            If true returns a dict of the showable items. Defaults to False.
+        return_discrete: bool
+            If true returns a dict of the showable items. Defaults to False.
```

Comment:
Return dict of gustaf discrete objects, for example,
      {Vertices, Edges, Faces}, instead of opening a window
      
Can apparently not suggest an inline change can one of you do this?